### PR TITLE
feat(router): BREAKING - router.push now uses PUSH navigation

### LINKED ISSUE #741: expo router navigation change w/ query parameters breaks browser history stack
### Which package manager are you using? (Yarn is recommended)

yarn

### Summary

Hi, I am trying to move through a series of modals that are available via query parameters. The flow is like so:

```
?modal=register_user&step=1
?modal=register_user&step=2
?modal=register_user&step=3
```

I'm using expo-router's `push` and/or `setParams` to modify these query parameters. I notice though, using these methods breaks the history's navigation state. The user cannot navigate back to a previous step i.e. `step=2` to `step=1` by using the browser back button. It seems that expo router preserves browser history state correctly when navigating between actual route paths, but not between query parameters.

I have attached a minimum example below, thank you. (Neither the Pressable or Link techniques work properly).

### Minimal reproducible example

To reproduce this error, click either the `<Link>` or the `<Pressable>` and then try to navigate back in the browser; the query parameters are not cleared until the back button is pressed 2-3 times, and it doesn't navigate to the current `pathname` it navigates to a prior screen (the URL that you were at before you entered this component—really wonky!).

```
import { P } from "@expo/html-elements";
import {
  Link,
  useGlobalSearchParams,
  usePathname,
  useRouter,
} from "expo-router";
import { Pressable, View } from "react-native";

export default () => {
  const { animal } = useGlobalSearchParams();
  const pathname = usePathname();
  const { push } = useRouter();
  const onPress = () => {
    push({
      pathname,
      params: {
        animal: "dog",
      },
    });
  };
  return (
    <View>
      {animal == "dog" ? <P>🐶</P> : <P>No Dogs here</P>}
      <View>
        <Link href="/test?animal=dog">See Dogs</Link>
        <Pressable onPress={onPress}>See Dogs (onPress)</Pressable>
      </View>
    </View>
  );
};

```

### LINKED ISSUE #758: Impossible to have a navigate to a route multiple times
### Which package manager are you using? (Yarn is recommended)

npm

### Summary

Hi,

Thank you so much for your job and expo router!

I tried expo router and I have an issue: I can't navigate to a route multiple times as react navigation ([Navigate to a route multiple times](https://reactnavigation.org/docs/navigating/#navigate-to-a-route-multiple-times)):

stack-push

### Minimal reproducible example

I tried with this code, but not work:

```
import { View, Text, Button } from 'react-native'
import React from 'react'
import { useRouter } from 'expo-router';

export default function settings() {
  const router = useRouter();

  return (
    <View>
      <Text>settings</Text>

      <Button title="Setings again" onPress={() => router.push('/home/settings')} />
    </View>
  )
}
```

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

`router.push()` should always most history forwards. However we are currently using the `NAVIGATE` action which may move history backwards if the screen is in history and the closest navigator is a Stack.

Adds `router.navigate` for when users wish to opt out of this behaviour.

Fix: https://github.com/expo/router/issues/758
Fix: https://github.com/expo/router/issues/741

Keeping this in draft until https://github.com/expo/expo/pull/24598 is merged

# How

This changes `router.push()` to a PUSH action if the lowest nested navigator is a stack and shared between the old & new routes. Otherwise it defaults back to NAVIGATE.

# Test Plan

Included test case

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by marklawlor** [comment]
Change name to `pushOrPop`

File: packages/expo-router/src/types.ts [hunk 1]
```diff
@@ -16,8 +16,10 @@ export interface RequireContext {
 export type PickPartial<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
 
 export type Router = {
-  /** Navigate to the provided href. */
+  /** Navigate to the provided href using a push operation if possible. */
   push: (href: Href) => void;
+  /** Navigate to the provided href. */
```
Comment (EvanBacon) [inline packages/expo-router/src/types.ts:21]: Make sure to copy these to the type generation file.

File: packages/expo-router/CHANGELOG.md [hunk 2]
```diff
@@ -6,10 +6,13 @@
 
 ### 🎉 New features
 
+- Add `router.navigate` and `navigate` to `useRouter` ([#24600](https://github.com/expo/expo/pull/24600) by [@marklawlor](https://github.com/marklawlor))
```
Comment (EvanBacon) [inline packages/expo-router/CHANGELOG.md:9]: Rename to something like `pushOrPop` to show that it's a superset of the push behavior.

**Comment by marklawlor** [comment]
Add navigation event types (replace/push/pushOrPop) to `<Link />`

**Comment by hasanaktas** [comment]
Is there an estimated time for release?

File: packages/expo-router/src/__tests__/navigation.test.tsx [hunk 3]
```diff
@@ -691,19 +691,15 @@ it('can push nested stacks with initial route names without creating circular re
   expect(screen).toHavePathname('/menu/123');
 });
 
-it('can push & replace with nested Slots', async () => {
+it.only('can push & replace with nested Slots', async () => {
```
Comment (marklawlor) [inline packages/expo-router/src/__tests__/navigation.test.tsx:694]: remove only

File: packages/expo-router/src/__tests__/smoke.test.tsx [hunk 4]
```diff
@@ -124,7 +124,7 @@ it('nested layouts', async () => {
   expect(HomeNested).toHaveBeenCalledTimes(1);
 });
 
-it('deep linking nested groups', async () => {
+it.only('deep linking nested groups', async () => {
```
Comment (marklawlor) [inline packages/expo-router/src/__tests__/smoke.test.tsx:127]: remove only

File: packages/@expo/cli/src/start/server/type-generation/routes.ts [hunk 5]
```diff
@@ -477,6 +477,10 @@ declare module "expo-router" {
     push: <T>(href: Href<T>) => void;
     /** Navigate to route without appending to the history. */
     replace: <T>(href: Href<T>) => void;
+    /** Navigate to the provided href. */
+    pushOrPop: <T>(href: Href<T>) => void;
+    /** If there's history that supports invoking the \`back\` function. */
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/routes.ts:482]: Make sure these match the comments in source.

File: packages/expo-router/src/types.ts (hunk 1 above)
Comment (EvanBacon) [inline packages/expo-router/src/types.ts:21]: Make this comment more explicit, indicating that it will always push, and that `pushOrPop` may be desired.

**Review by EvanBacon — CHANGES_REQUESTED** [review]
Update the docs too and add a guide about differences between push / pushOrPop.

**Comment by marklawlor** [comment]
Renamed back to `navigate`

Now `Link` uses `NAVIGATE` by default and `router.push` uses `PUSH`.

You can add the `push` prop to `Link` for it to use the `PUSH` behaviour
You can use `router.navigate` for it to use the `NAVIGATE` behaviour

**Comment by marklawlor** [comment]
Type code I'm going to defer and add to this PR https://github.com/expo/expo/pull/25654

File: packages/expo-router/src/global-state/router-store.tsx [hunk 6]
```diff
@@ -101,7 +102,7 @@ export class RouterStore {
     }
 
     /**
-     * Counter intuitively - this fires AFTER both React Navigations state change and the subsequent paint.
+     * Counter intuitively - this fires AFTER both React Navigation state change and the subsequent paint.
```
Comment (EvanBacon) [inline packages/expo-router/src/global-state/router-store.tsx:105]:
```suggestion
     * Counter intuitively - this fires AFTER both React Navigation's state changes and the subsequent paint.
```
haha

File: docs/pages/router/navigating-pages.mdx [hunk 7]
```diff
@@ -104,6 +116,29 @@ export default function Page() {
 }
 ```
 
+## Pushing screens
+
+By default, links "push" new routes onto the navigation stack or unwind to a existing route. You can use the `push` prop to always push the route onto the stack.
```
Comment (EvanBacon) [inline docs/pages/router/navigating-pages.mdx:121]:
```suggestion
By default, links `navigate` to the nearest route in the navigation stack, either by pushing a new route or unwinding to an existing route. You can use the `push` prop to always push the route onto the stack.
```

**Review by EvanBacon — APPROVED** [review]
lgtm, pending docs pass

File: docs/pages/router/navigating-pages.mdx [hunk 8]
```diff
@@ -55,6 +55,17 @@ export default function Page() {
 }
 ```
 
+## Understanding native navigation
+
+Expo Router employs a stack-based navigation approach. This means that each new route you navigate to is added to a stack. If you navigate to a route that's already in the stack, the stack is unwound back to that existing route.
```
Comment (amandeepmittal) [inline docs/pages/router/navigating-pages.mdx:60]:
```suggestion
Expo Router uses a stack-based navigation approach. Each new route you navigate to gets added to a stack. If you navigate a route already in the stack, the stack unwinds back to that existing route.
```

- Let's keep this in present tense for clairty
- Also, let's be a bit concise

File: docs/pages/router/navigating-pages.mdx [hunk 9]
```diff
@@ -55,6 +55,17 @@ export default function Page() {
...
+
+For example, if you navigate from `/feed` to `/profile`, the stack will contain `/feed` and `/profile`. If you then navigate to `/settings`, the stack will contain `/feed`, `/profile`, and `/settings`. If you then navigate to `/feed`, the stack will unwind back to `/feed`.
```
Comment (amandeepmittal) [inline docs/pages/router/navigating-pages.mdx:62]:
```suggestion
For example, when you navigate from `/feed` to `/profile`, the stack contains `/feed` and `/profile`. If you then navigate to `/settings`, the stack contains `/feed`, `/profile`, and `/settings`. If you then navigate back to `/feed`, the stack unwinds back to `/feed`.
```

File: docs/pages/router/navigating-pages.mdx [hunk 10]
```diff
@@ -55,6 +55,17 @@ export default function Page() {
...
+
+If you wish to navigate to a route without the stack unwinding, you can use the `push` prop on the `<Link>` component. This will always push the route onto the stack, even if it already exists.
```
Comment (amandeepmittal) [inline docs/pages/router/navigating-pages.mdx:64]:
```suggestion
To navigate to a route without the stack unwinding, you can use the `push` prop on the `<Link>` component. This alwayss pushes the route onto the stack, even if it already exists.
```

File: docs/pages/router/navigating-pages.mdx [hunk 11]
```diff
@@ -55,6 +55,17 @@ export default function Page() {
...
+
+In contrast, the `replace` method substitutes the current route in the navigation stack with a new one, effectively replacing the current screen with the new one without adding to the stack.
+
+
```
Comment (amandeepmittal) [inline docs/pages/router/navigating-pages.mdx:68]:
```suggestion
```

File: docs/pages/router/navigating-pages.mdx [hunk 12]
```diff
@@ -71,8 +82,9 @@ export function logout() {
 
 The `router` object is immutable and contains the following functions:
 
-- **push**: `(href: Href) => void` Navigate to a route. You can provide a full path like **/profile/settings** or a relative path like **../settings**. Navigate to dynamic routes by passing an object like `{ pathname: 'profile', params: { id: '123' } }`.
-- **replace**: `(href: Href) => void` Same API as push but replaces the current route in the history instead of pushing a new one. This is useful for redirects.
+- **naviate**: `(href: Href) => void` Navigate to a route using the default behaviour. You can provide a full path like **/profile/settings** or a relative path like **../settings**. Navigate to dynamic routes by passing an object like `{ pathname: 'profile', params: { id: '123' } }`.
```
Comment (amandeepmittal) [inline docs/pages/router/navigating-pages.mdx:85]:
```suggestion
- **navigate**: `(href: Href) => void` Navigate to a route using the default behavior. You can provide a full path like **/profile/settings** or a relative path like **../settings**. Navigate to dynamic routes by passing an object like `{ pathname: 'profile', params: { id: '123' } }`.
```

- Typo and we prefer American English for spellings

**Review by amandeepmittal — APPROVED** [review]
Minor suggestions but lgtm 👍

File: docs/pages/router/navigating-pages.mdx [hunk 13]
```diff
@@ -104,6 +115,29 @@ export default function Page() {
 }
 ```
 
+## Pushing screens
+
+By default, links `navigate` to the nearest route in the navigation stack, either by pushing a new route or unwinding to an existing route. You can use the `push` prop to always push the route onto the stack.
+
+{/* prettier-ignore */}
+```js app/index.js
```
Comment (Simek) [inline docs/pages/router/navigating-pages.mdx:123]:
```suggestion
```jsx app/index.js
```

Syntax highlight nit.

**Comment by kennethstarkrl** [comment]
@marklawlor The push issue still exists for me after this change in v3.
Doing a `router.push()` to the same route with different params changes the url with the new params, but the page doesn't change.

I commented out the following from `/expo-router/build/global-state/routing.js` `getNavigateAction()` and now push is working and navigates as I expected to.
```js
if (route.state && routesAreEqual && currentRoute.state) {
    return getNavigateAction(route.state, currentRoute.state, type);
}
```

My guess is that since only the route params changed, all the routes and states are equal. Might need an additional check to see if the params are equal as well. 

I modified it adding a params are equal check  and it seems to be working
_(not entirely sure if this is optimal or correct)_
```js
function getNavigateAction(state, parentState, type = 'NAVIGATE') {

    const route = state.routes[state.routes.length - 1];

    const currentRoute = parentState.routes.find((parentRoute) => parentRoute.name === route.name);
    const routesAreEqual = parentState.routes[parentState.index] === currentRoute;
    const paramsAreEqual = rewriteNavigationStateToParams(parentState).params === rewriteNavigationStateToParams(state).params;
    // If there is nested state and the routes are equal, we should keep going down the tree
    if (route.state && routesAreEqual && currentRoute.state && paramsAreEqual) {
        return getNavigateAction(route.state, currentRoute.state, type);
    }
    // Either we reached the bottom of the state or the point where the routes diverged
    const { screen, params } = rewriteNavigationStateToParams(state);
    if (type === 'PUSH' && parentState.type !== 'stack') {
        type = 'NAVIGATE';
    }
    else if (type === 'REPLACE' && parentState.type === 'tab') {
        type = 'JUMP_TO';
    }
    return {
        type,
        target: parentState.key,
        payload: {
            name: screen,
            params,
        },
    };
}
```

**Pull request merged**